### PR TITLE
Improve deprecation checks for smoke tests

### LINKED ISSUE #15983: Ensure deprecation warnings in smoke tests are followed up on
When a deprecation warning is added, we currently have the infrastructure to notice if it appears in one of our smoke tests; the smoke test will fail without explicitly adding the deprecation message as expected.

It would be good to make sure the person adding the new deprecation has a plan to follow up with the maintained of the plugin to eliminate the deprecation before the next major release of Gradle.

It would also be good to have a way to look up how these follow ups are going from time to time, maybe after each minor release of Gradle.

A few changes we could do to facilitate this:

1) Make the expectation of no deprecation warnings the default. Currently we still need to add `expectNoDeprecationWarnings(result)` in a smoke test to enforce the check for deprecation warnings, but this should be on by default.

2) Replace `expectDeprecationWarnings()` with individual calls to `expectDeprecationWarning()` and for each individual expected deprecation require a second parameter with a link to an issue of PR.

3) Add `expectLegacyDeprecationWarning()` without an issue/PR link. While it is essential for the latest version of a plugin to work without deprecation warnings, we also have smoke tests for earlier versions of a plugin. With those versions it might be acceptable to have some deprecation warnings that are not going to be removed.

4) For the external plugin validation smoke tests, change `failsWith()` to take a single warning/error at once, along with `legacyVersionFailsWith()` that does not require a link.

---
cc: @gradle/execution

### PR DESCRIPTION
**Pull request by lptr**
This PR does a few things:

- Makes it the default to always check for deprecation warnings in smoke tests. Previously it was possible to not call either `expectDeprecationWarnings()` nor `expectNoDeprecationWarnings()` and get any deprecation warnings ignored accidentally. The old methods are gone, the new API looks like this:
    ```groovy
        when:
        runner('assembleDebug')
            .expectDeprecationWarning("The JavaExecHandleBuilder.setMain(String) method has been deprecated. " +
                "This is scheduled to be removed in Gradle 8.0. Please use the mainClass property instead. " +
                "Consult the upgrading guide for further information: https://docs.gradle.org/${GradleVersion.current().version}/userguide/upgrading_version_7.html#java_exec_properties",
                "https://github.com/third-party/gradle-plugin/issues/456") // <-- followup link
            .build()
    ```
- Adds a mandatory followup link to each expected warning. The workflow is as follows:
    1) you add a new deprecation in Gradle,
    2) it fails a smoke test, so you contact the owner of the respective plugin, and file an issue,
    3) until the issue is fixed, you add the expectation to the smoke test, and link to the issue you created.
    Having the parameter serves as a reminder to not forget filing the issue.
- The PR adds some previously missing expectations for deprecations.
- It also removes some assertions related to deprecations in old (<4.0) AGP versions that are not used anymore.

Fixes #15983

### CODE REVIEW

**Comment by lptr** [comment]
@bot-gradle check smoke.

**Comment by lptr** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge.

**Comment by lptr** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge.

**Comment by lptr** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge.

**Comment by lptr** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge.

**Comment by lptr** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge.

**Comment by lptr** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge.

**Comment by lptr** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge.

**Comment by lptr** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge.

**Comment by lptr** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge.

**Comment by lptr** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge.

**Comment by lptr** [comment]
OK, I've already triggered AllSmokeTestsReadyForMerge build for you.

**Comment by lptr** [comment]
@bot-gradle test AllSmokeTestsReadyForMerge.

**Review by wolfs — APPROVED** [review]
LGTM!

**Comment by lptr** [comment]
@bot-gradle test and merge please.

**Comment by lptr** [comment]
@bot-gradle test and merge please.

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidPluginsSmokeTest.groovy [hunk 1]
```diff
@@ -84,45 +82,21 @@ class AndroidPluginsSmokeTest extends AbstractPluginValidatingSmokeTest implemen
         ))
 
         when: 'first build'
-        def result = runner.build()
+        def result = runner
+            .expectDeprecationWarningIf(agpVersion.startsWith('4.0.2') || agpVersion.startsWith('4.1'),
+                "The WorkerExecutor.submit() method has been deprecated. " +
+                    "This is scheduled to be removed in Gradle 8.0. Please use the noIsolation(), classLoaderIsolation() or processIsolation() method instead. " +
+                    "See https://docs.gradle.org/${GradleVersion.current().version}/userguide/upgrading_version_5.html#method_workerexecutor_submit_is_deprecated for more details.",
+                "TODO Add followup issue")
```
Comment (rieske) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidPluginsSmokeTest.groovy:90]: `expectLegacyDeprecationWarning` here? This deprecation is fixed in newer versions of the plugin.

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AsciidoctorPluginSmokeTest.groovy [hunk 2]
```diff
@@ -44,7 +45,12 @@ class AsciidoctorPluginSmokeTest extends AbstractPluginValidatingSmokeTest {
             """.stripIndent()
 
         when:
-        runner('asciidoc').build()
+        runner('asciidoc')
+            .expectDeprecationWarning("The JavaExecHandleBuilder.setMain(String) method has been deprecated. " +
+                "This is scheduled to be removed in Gradle 8.0. " +
+                "Please use the mainClass property instead. Consult the upgrading guide for further information: https://docs.gradle.org/${GradleVersion.current().version}/userguide/upgrading_version_7.html#java_exec_properties",
+                "TODO Add followup issue")
```
Comment (rieske) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AsciidoctorPluginSmokeTest.groovy:52]: https://github.com/asciidoctor/asciidoctor-gradle-plugin/issues/602

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/KotlinPluginSmokeTest.groovy [hunk 3]
```diff
@@ -44,22 +43,19 @@ class KotlinPluginSmokeTest extends AbstractPluginValidatingSmokeTest implements
         replaceVariablesInBuildFile(kotlinVersion: version)
 
         when:
-        def result = build(workers, 'run')
+        def result = runner(workers, 'run')
+            .expectDeprecationWarningIf(workers,
+                "The WorkerExecutor.submit() method has been deprecated. " +
+                    "This is scheduled to be removed in Gradle 8.0. Please use the noIsolation(), classLoaderIsolation() or processIsolation() method instead. " +
+                    "See https://docs.gradle.org/${GradleVersion.current().version}/userguide/upgrading_version_5.html#method_workerexecutor_submit_is_deprecated for more details.",
+                "TODO Add followup issue"
```
Comment (rieske) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/KotlinPluginSmokeTest.groovy:51]: https://youtrack.jetbrains.com/issue/KT-46019

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/KotlinPluginSmokeTest.groovy [hunk 4]
```diff
@@ -83,7 +79,14 @@ class KotlinPluginSmokeTest extends AbstractPluginValidatingSmokeTest implements
         replaceVariablesInBuildFile(kotlinVersion: version)
 
         when:
-        def result = build(workers, 'compileKotlin2Js')
+        def result = runner(workers, 'compileKotlin2Js')
+            .expectDeprecationWarningIf(workers,
+                "The WorkerExecutor.submit() method has been deprecated. " +
+                    "This is scheduled to be removed in Gradle 8.0. Please use the noIsolation(), classLoaderIsolation() or processIsolation() method instead. " +
+                    "See https://docs.gradle.org/${GradleVersion.current().version}/userguide/upgrading_version_5.html#method_workerexecutor_submit_is_deprecated for more details.",
+                "TODO Add followup issue"
```
Comment (rieske) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/KotlinPluginSmokeTest.groovy:87]: https://youtrack.jetbrains.com/issue/KT-46019

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidPluginsSmokeTest.groovy (hunk 1 above)
Comment (lptr) [reply subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidPluginsSmokeTest.groovy:90]: Good point.

**Comment by lptr** [comment]
@bot-gradle test and merge.

**Comment by rieske** [comment]
@bot-gradle test and merge

**Comment by rieske** [comment]
@bot-gradle test and merge

**Comment by rieske** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Comment by rieske** [comment]
@bot-gradle test and merge

**Comment by rieske** [comment]
@bot-gradle test and merge please please please

**Pull request merged**